packages/kubernetes-common/src/types.ts

```diff
@@ -0,0 +1,208 @@
+}
+
+export interface KubernetesRequestBody {
+  auth?: {
+    google?: string;
+  };
+  entity: Entity;
+}
+
+export interface ClusterObjects {
+  cluster: { name: string };
+  resources: FetchResponse[];
+  errors: KubernetesFetchError[];
+}
+
+export interface ObjectsByEntityResponse {
+  items: ClusterObjects[];
+}
+
+export interface FetchResponseWrapper {
+  errors: KubernetesFetchError[];
+  responses: FetchResponse[];
+}
+
+export type FetchResponse =
+  | PodFetchResponse
+  | ServiceFetchResponse
+  | ConfigMapFetchResponse
+  | DeploymentFetchResponse
+  | ReplicaSetsFetchResponse
+  | HorizontalPodAutoscalersFetchResponse
+  | IngressesFetchResponse
+  | CustomResourceFetchResponse;
+
+// TODO fairly sure there's a easier way to do this
+
+export type KubernetesObjectTypes =
+  | 'pods'
+  | 'services'
+  | 'configmaps'
+  | 'deployments'
+  | 'replicasets'
+  | 'horizontalpodautoscalers'
+  | 'ingresses'
+  | 'customresources';
+
+export interface PodFetchResponse {
+  type: 'pods';
+  resources: Array<V1Pod>;
+}
+
+export interface ServiceFetchResponse {
+  type: 'services';
+  resources: Array<V1Service>;
+}
+
+export interface ConfigMapFetchResponse {
+  type: 'configmaps';
+  resources: Array<V1ConfigMap>;
+}
+
+export interface DeploymentFetchResponse {
+  type: 'deployments';
+  resources: Array<V1Deployment>;
+}
+
+export interface ReplicaSetsFetchResponse {
+  type: 'replicasets';
+  resources: Array<V1ReplicaSet>;
+}
+
+export interface HorizontalPodAutoscalersFetchResponse {
+  type: 'horizontalpodautoscalers';
+  resources: Array<V1HorizontalPodAutoscaler>;
+}
+
+export interface IngressesFetchResponse {
+  type: 'ingresses';
+  resources: Array<ExtensionsV1beta1Ingress>;
+}
+
+export interface CustomResourceFetchResponse {
+  type: 'customresources';
+  resources: Array<any>;
+}
+
+export interface ObjectFetchParams {
+  serviceId: string;
+  clusterDetails: ClusterDetails;
+  objectTypesToFetch: Set<KubernetesObjectTypes>;
+  labelSelector: string;
+  customResources: CustomResource[];
+}
+
+// Fetches information from a kubernetes cluster using the cluster details object
+// to target a specific cluster
+export interface KubernetesFetcher {
+  fetchObjectsForService(
+    params: ObjectFetchParams,
+  ): Promise<FetchResponseWrapper>;
+}
+
+// Used to locate which cluster(s) a service is running on
+export interface KubernetesServiceLocator {
+  getClustersByServiceId(serviceId: string): Promise<ClusterDetails[]>;
+}
+
+// Used to load cluster details from different sources
+export interface KubernetesClustersSupplier {
+  getClusters(): Promise<ClusterDetails[]>;
+}
+
+export type KubernetesErrorTypes =
+  | 'BAD_REQUEST'
+  | 'UNAUTHORIZED_ERROR'
+  | 'SYSTEM_ERROR'
+  | 'UNKNOWN_ERROR';
+
+export interface KubernetesFetchError {
+  errorType: KubernetesErrorTypes;
+  statusCode?: number;
+  resourcePath?: string;
+}
+
+export interface ConfigClusterLocatorMethod {
+  /**
+   * @visibility frontend
+   */
```

Comment:
Noticed these here which seems to be a trick to re-use types between config and the backend implementation. It totally doesn't work for published packages though, and will lead to some bad habits when reading values from config. Seeing as this PR is gonna break that bit even more I think we could take care of it here :grin:

Reply:
Oh, gotcha.

Reply:
oh, actually... now I'm a bit confused.

The schema.d.ts is in plugins/kubernetes and it refers to ClusterLocatorMethod (which is either ConfigClusterLocatorMethod or GKEClusterLocatorMethod). First, of all, does it make sense to have this config in the frontend package? I don't see this being used there at all.

I feel I should move the shema.d.ts to the backend and all associated types and export 'em from there (all the deps of `Config`). WDYS?

Reply:
It's on both actually.

Reply:
So I spoke with Ben & Matthew and added a commit removing the schema from kubernetes. Only the types that are actually shared are now in common, the ones used only in the backend rest there. Type surgery.